test: Rich text test helpers mimic user events

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Refactor the existing native mobile rich text test helpers to mimic user events,
e.g. typing within an input, and add a convenience helper for selecting rich
text.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Test helpers that more closely mimic user behavior will produce tests that are
more sound. Additionally, the simplified test helpers will increase the
comprehensibility of our tests.

This was originally discussed in https://github.com/WordPress/gutenberg/pull/49732#discussion_r1165496052.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- `changeAndSelectTextOfRichText` defaults cursor to end of new text.
- Consolidate `changeTextOfRichText` and `changeAndSelectTextOfRichText` into
  `typeInRichText`.
- Add `selectRangeInRichText` helper.
- Simplify existing tests with new helpers.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
n/a

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
n/a

## Screenshots or screencast <!-- if applicable -->
n/a

dcalhoun committed:
test: changeAndSelectTextOfRichText defaults cursor to end of new text

This will mimic natural typing where the cursor is placed at the end of
newly typed text by default.

dcalhoun committed:
test: Consolidate multiple Rich Text change text helpers

The existing helpers contained substantial overlap and were coupled to
implementation details of the `RichText` event handlers. This refactor
creates a new, single helper to mimic typing into rich text fields as a
user might.

dcalhoun committed:
test: Add selectTextInRichText helper

Simply process for selecting rich text values to improve test
comprehensibility.

dcalhoun committed:
test: Simplify tests with new helpers

It is no longer necessary to pass the length of the test strings as the
new helper automatically tracks the string length to place the cursor at
the end of the new value, mimicking a user typing.

dcalhoun commented on packages/block-library/src/list/test/edit.native.js:
```diff
@@ -347,10 +347,7 @@ describe( 'List block', () => {
-			initialSelectionStart: 0,
-			initialSelectionEnd: 3,
-		} );
+		selectTextInRichText( listItemField );
```
This specific example relies upon `selectTextInRichText`'s default parameters of `selectionStart = 0` and `selectionEnd = selectionStart`. It is convenient to not need to pass `0` here to set the cursor to the beginning of the input. 

That said, does the name `selectTextInRichText` too strongly imply that (all) text will be selected by default? We cannot select _all_ text by default, as the helper is not receiving or know the string length. However, we could make `selectionStart` (and possibly `selectionEnd`) required to avoid any ambiguity of how this helper functions. 

What is better for this helper: convenience or clarity?

fluiddot replied:
> That said, does the name `selectTextInRichText` too strongly imply that (all) text will be selected by default? We cannot select _all_ text by default, as the helper is not receiving or know the string length.

In `typeInRichText` function we're getting the text length by checking the `value` prop of `RichText`, so maybe we could do the same for selection.

> However, we could make `selectionStart` (and possibly `selectionEnd`) required to avoid any ambiguity of how this helper functions.

Yep, when I read `selectTextInRichText` my first impression was that it will select all text. I was thinking about naming it differently like `moveCaretInRichtText`, but not sure if the name implies text selection. As you mentioned, probably making the argument `selectionStart` required would help to improve clarity. Regarding `selectionEnd`, not sure if I'd also make it required. 

> What is better for this helper: convenience or clarity?

Good question. I'd advocate for clarity in the spirit of making the tests easier to read. WDYT?

dcalhoun replied:
> In `typeInRichText` function we're getting the text length by checking the `value` prop of `RichText`, so maybe we could do the same for selection.

Correct, but the `value` is unavailable within the `selectTextInRichText` function scope. We would have to add a special parameter to `typeInRichText` allowing communication of a "`typeInRichText` select the entire value (that only you know about) for me" intent. I'm less inclined expand the API surface of `typeInRichText` for this one-off, hidden (internal to the `selectTextInRichText` helper) use case. 

```js
function selectTextInRichText(richText, start, end = start) {
  typeInRichText(richText, "", { selectAll: true });
}
```

fluiddot commented on test/native/integration-test-helpers/rich-text-select.js:
```diff
@@ -0,0 +1,20 @@
+ */
+export const selectTextInRichText = (
+	richText,
+	selectionStart = 0,
```
As mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/49804/files#r1165601733), probably we could make this argument required.

fluiddot commented on test/native/integration-test-helpers/rich-text-type.js:
```diff
@@ -17,29 +17,26 @@ function insertTextAtPosition( text, newText, start, end ) {
- * @param {number}                                          [options.selectionEnd]          Selection end position.
+ * @param {number}                                          [options.initialSelectionStart] Selection start position before the text is inserted.
+ * @param {number}                                          [options.initialSelectionEnd]   Selection end position before the text is inserted.
+ * @param {number}                                          [options.selectionStart]        Selection start position after the text is inserted.
+ * @param {number}                                          [options.selectionEnd]          Selection end position after the text is inserted.
```
For clarity, I'm thinking about renaming `selectionStart` and `selectionEnd` to `finalSelectionStart` and `finalSelectionEnd` as these are the final selection values after the text is inserted, WDYT?

dcalhoun committed:
test: Refactor selectTextInRichText

Rename helper and require range start position to improve clarity of the
purpose and function of the now `selectRangeInRichText` helper.

dcalhoun committed:
test: Rename typeInRichText selection parameters

Improve clarity by further differentiating the initial and final
selection range.

dcalhoun commented on test/native/integration-test-helpers/rich-text-select-range.js:
```diff
@@ -0,0 +1,23 @@
+ * @param {number}                                          end      Selection end position.
+ *
+ */
+export const selectRangeInRichText = ( richText, start, end = start ) => {
+	if ( typeof start !== 'number' ) {
+		throw new Error( 'A numerical range start value must be provided.' );
+	}
```
In an effort to [mitigate any confusion](https://github.com/WordPress/gutenberg/pull/49804#discussion_r1165601733) as to how this helper functions, I renamed it from `selectTextInRichText` to `selectRangeInRichText`. I also made the `start` parameter required.

fluiddot replied:
Sounds great, the name is clearly descriptive to avoid confusion 👍 .

fluiddot reviewed (approved):
LGTM 🎊 ! Thanks @dcalhoun for improving the test helpers, this is a great enhancement that will help us writing easier next integration tests 🙇 .
